components/salsa-2022-macros/src/tracked_fn.rs

```diff
@@ -85,6 +71,206 @@ impl crate::options::AllowedOptions for TrackedFn {
     const CONSTRUCTOR_NAME: bool = false;
 }
 
+type ImplArgs = Options<TrackedImpl>;
+
+pub(crate) fn tracked_impl(
+    args: proc_macro::TokenStream,
+    mut item_impl: syn::ItemImpl,
+) -> syn::Result<TokenStream> {
+    let args: ImplArgs = syn::parse(args)?;
+    let self_type = match &*item_impl.self_ty {
```

Comment:
Do we have to check for trait impls?

e.g.,

```rust
#[salsa::tracked]
impl Foo for MyType {
}
```

Reply:
I probably should do but I'm not sure what your suggesting, with this change trait impls kinda just work, the methods must take (self, &dyn Db, ...) and it doesn't include the trait name in the generated struct so you can still only have one method with that name. The two options I see:
1. Accept trait impls but include the trait name in the generaated struct to reduce conflicts
2. Ban trait impls to keep things simple until we find a reason that they're needed

Reply:
Hmm! Interesting. It hadn't occurred to me that they might just work but... I guess ... they probably do, so long as they take a `&dyn crate::Db` as argument? 

I don't see any reason to make them an error, in that case, but can we add a test that shows they work?

Reply:
I don't have a strong opinion about the name of the generated type, but including the trait name seems like a good idea.

Reply:
Maybe `Type_Trait_method_name` ?

Reply:
(request: add a test showing that trait impls work, adjust name of generated type)

Reply:
Done. Although I did notice this also works:
```rust
#[salsa::jar(db = Db)]
struct Jar(
    MyInput,
    MyInput_TrackedTrait_tracked_trait_fn,
);

trait TrackedTrait<T> {
    fn tracked_trait_fn(self, db: &dyn Db, val: T) -> T;
}

#[salsa::tracked]
impl TrackedTrait<u32> for MyInput {
    #[salsa::tracked]
    fn tracked_trait_fn(self, db: &dyn Db, val: u32) -> u32 {
        self.field(db) * val
    }
}
```
I think it's fine but could require breaking changes in the future if we want to support multiple generic impls in the same module in the future.